libs/s25main/ingameWindows/iwEconomicProgress.cpp

```diff
@@ -0,0 +1,200 @@
+    const std::vector<EconomyModeHandler::econTeam>& teams = gwv.GetWorld().econHandler->GetTeams();
+    for(unsigned int t = 0; t < teams.size(); t++)
+    {
+        DrawPoint drawPt = GetDrawPos() + DrawPoint(37 + (t + 1) * 63, 22);
+        unsigned int height = 24;
+        unsigned int ystep = height / teams[teamorder[t]].num_players_in_team;
+        unsigned int ypos = 0;
+        for(unsigned i = 0; i < gwv.GetWorld().GetNumPlayers(); ++i)
+        {
+            if(teams[teamorder[t]].inTeam(i))
+            {
+                if(height - ypos < 2 * ystep)
+                    ystep = height - ypos;
+                DrawRectangle(Rect(drawPt + DrawPoint(0, ypos), Extent(62, ystep)), gwv.GetWorld().GetPlayer(i).color);
+                ypos += ystep;
+            }
+        }
+    }
+}
+
+void iwEconomicProgress::Msg_ButtonClick(const unsigned ctrl_id)
+{
+    switch(ctrl_id)
+    {
+        break;
+        case 25: // Hilfe
+        {
+            WINDOWMANAGER.ReplaceWindow(
+              std::make_unique<iwHelp>(_("This window shows the wares that should be collected for "
+                                         "the economy mode. Displayed is the current inventory of "
+                                         "the player, his team and after game end of the opponent "
+                                         "team.")));
+        }
+        break;
+    }
+}
+
+void iwEconomicProgress::Msg_PaintBefore()
+{
+    IngameWindow::Msg_PaintBefore();
+
+    const GameWorldBase& world = gwv.GetWorld();
+    EconomyModeHandler* eH = world.econHandler;
+    const std::vector<EconomyModeHandler::econTeam>& teams = eH->GetTeams();
+    eH->UpdateAmounts(); // make sure the amounts are current
+    const GamePlayer& mainPlayer = gwv.GetPlayer();
+
+    for(unsigned int i = 0; i < eH->numGoodTypesToCollect; i++)
+    {
+        for(unsigned j = 0; j < 1 + teams.size(); j++)
+        {
+            auto* text = GetCtrl<ctrlTextDeepening>(300 + 10 * j + i);
+            if(j == 0)
+            {
+                text->SetText(std::to_string(eH->GetAmount(i, mainPlayer.GetPlayerId())));
+            } else if(j == 1 || eH->isOver())
+            {
+                text->SetText(std::to_string(teams[teamorder[j - 1]].teamAmounts[i]));
+
+                // White color at game end to mark good types that the team has won
+                if(eH->isOver() && teams[teamorder[j - 1]].teamAmounts[i] >= eH->GetMaxTeamAmount(i))
+                {
+                    text->SetTextColor(COLOR_WHITE);
+                } else
+                {
+                    if(j == 1)
+                        text->SetTextColor(COLOR_YELLOW);
+                    else
+                        text->SetTextColor(COLOR_RED);
+                }
+            } else
+            {
+                text->SetText("???");
+            }
+        }
+    }
+
+    // Refresh game progress tracker
+    unsigned elapsed = 100;
+    if(!eH->isOver() && eH->GetEndFrame() > 0)
```

Comment:
Why is the 2nd condition required? Could there be a better named function? E.g. `isInfiniteDuration` or so?